e3sm_diags/driver/zonal_mean_2d_driver.py

```diff
@@ -1,283 +1,347 @@
-                ):
-                    plevs = default_plevs
-                logger.info(f"Selected pressure level: {plevs}")
-
-                mv1_p = utils.general.convert_to_pressure_levels(
-                    mv1, plevs, test_data, var, season
-                )
-                mv2_p = utils.general.convert_to_pressure_levels(
-                    mv2, plevs, ref_data, var, season
-                )
+    test_ds = Dataset(parameter, data_type="test")
+    ref_ds = Dataset(parameter, data_type="ref")
 
-                # Note this is a special case to handle small values of stratosphere specific humidity.
-                # The general derived variable process converts specific humidity to units [g/kg]
-                # Following converts from g/kg to ppm
-
-                if (
-                    parameter.current_set == "zonal_mean_2d_stratosphere"
-                    and parameter.var_id == "Q"
-                ):
-                    mv1_p = mv1_p * 1000.0
-                    mv1_p.units = "ppm"
-                    mv2_p = mv2_p * 1000.0
-                    mv2_p.units = "ppm"
-                # Regrid towards the lower resolution of the two
-                # variables for calculating the difference.
-                mv1_p_reg, mv2_p_reg = utils.general.regrid_to_lower_res(
-                    mv1_p,
-                    mv2_p,
-                    parameter.regrid_tool,
-                    parameter.regrid_method,
-                )
-
-                diff_p = mv1_p_reg - mv2_p_reg
-                diff = cdutil.averager(diff_p, axis="x")
+    for var_key in variables:
+        logger.info("Variable: {}".format(var_key))
+        parameter.var_id = var_key
 
-                mv1_p = cdutil.averager(mv1_p, axis="x")
-                mv2_p = cdutil.averager(mv2_p, axis="x")
+        for season in seasons:
+            parameter._set_name_yrs_attrs(test_ds, ref_ds, season)
 
-                # Make sure mv1_p_reg and mv2_p_reg have same mask
-                mv1_p_reg = mv2_p_reg + diff_p
-                mv2_p_reg = mv1_p_reg - diff_p
+            ds_test = test_ds.get_climo_dataset(var_key, season)
+            ds_ref = ref_ds.get_ref_climo_dataset(var_key, season, ds_test)
 
-                mv1_reg = cdutil.averager(mv1_p_reg, axis="x")
-                mv2_reg = cdutil.averager(mv2_p_reg, axis="x")
+            # Store the variable's DataArray objects for reuse.
+            dv_test = ds_test[var_key]
+            dv_ref = ds_ref[var_key]
 
-                parameter.output_file = "-".join(
-                    [ref_name, var, season, parameter.regions[0]]
-                )
-                parameter.main_title = str(" ".join([var, season]))
+            is_vars_3d = has_z_axis(dv_test) and has_z_axis(dv_ref)
+            is_dims_diff = has_z_axis(dv_test) != has_z_axis(dv_ref)
 
-                if parameter.diff_type == "relative":
-                    diff = diff / mv2_reg * 100.0
+            if not is_vars_3d:
```

Comment:
The original codes are misleading...the zonal_mean_2d set should only apply to 3d datasets, so this https://github.com/E3SM-Project/e3sm_diags/pull/752#issuecomment-1928488372 code block should never been used. So `if not is _vars_3d` the input data should be invalid for this set.